### PR TITLE
Julien Chien

### PR DESCRIPTION
**Pull request by wholien**

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -8,97 +11,99 @@ pub struct DB<T> {
...
+    entries: Vec<&'a mut T>,
 }
 
 /// Filters a DBView using the the given predicate.
-pub fn filter_one<T, F>(view: &DBView<T>, predicate: F) -> DBView< T>
+pub fn filter_one<'a, T, F>(view: &DBView<'a, T>, predicate: F) -> DBView<'a, T>
     where F: Fn(&T) -> bool
 {
-    unimplemented!()
+  DBView{ entries: view.entries.clone().into_iter().filter(|&x| predicate(x)).collect() }
```
Comment (alex-ozdemir) [inline src/lib.rs:30]: Good use of iterator methods!

You could however, do better. You might imagine
- not cloning the initial vector, and instead 
- _only cloning the references you need to clone_, immediately before collecting them. Take a look at `Iterator::cloned`.

File: src/lib.rs [hunk 2]
```diff
@@ -8,97 +11,99 @@ pub struct DB<T> {
...
+pub fn filter_two<'a, 'b, T, F>(view_a: &DBView<'a, T>,
+                        view_b: &DBView<'b, T>,
                         predicate: F)
-                        -> (DBView<T>, DBView<T>)
+                        -> (DBView<'a, T>, DBView<'b, T>)
     where F: Fn(&T) -> bool
 {
-    unimplemented!()
+  (DBView{ entries: view_a.entries.clone().into_iter().filter(|&x| predicate(x)).collect() },
+  DBView{ entries: view_b.entries.clone().into_iter().filter(|&x| predicate(x)).collect() })
```
Comment (alex-ozdemir) [inline src/lib.rs:42]: Same

File: src/lib.rs [hunk 3]
```diff
@@ -8,97 +11,99 @@ pub struct DB<T> {
...
+        DB{ data: data }
     }
 
     /// Creates a new DBView containing all entries in `self` which satisfy `predicate`
-    pub fn select_where<F>(&self, predicate: F) -> DBView<T>
+    pub fn select_where<'a, F>(&'a self, predicate: F) -> DBView<'a, T>
         where F: Fn(&T) -> bool
     {
-        unimplemented!()
+        DBView{ entries: self.data.iter().filter(|i| predicate(i)).collect::<Vec<&'a T>>()}
```
Comment (alex-ozdemir) [inline src/lib.rs:55]: The typechecker should be able to infer how `collect` should be parameterized here.

**Review by alex-ozdemir — COMMENTED** [review]
Nice work!

Fairly idiomatic code (although you could avoid some unnecessary copies).

This is, however, late. You haven't used you euro yet, so consider it used here.

**Comment by alex-ozdemir** [comment]
Also, remember to fill out the post-assignment questionnaire, here:
https://docs.google.com/forms/d/e/1FAIpQLSfUWFMPPcQvf8MCA3pUG537nL0UGnzdme1bagT-yZUssdIXtQ/viewform

File: src/lib.rs (hunk 1 above)
Comment (wholien) [reply src/lib.rs:30]: so this line could then be:
`DBView{ entries: view.entries.into_iter().filter(|&x| predicate(x)).cloned().collect() }` ?

Comment (alex-ozdemir) [reply src/lib.rs:30]: Indeed!

**Pull request closed without merging**